[JIT] Improve error message for unsupported Optional types

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#44054 [JIT] Improve error message for unsupported Optional types**

**Summary**
This commit improves the error message that is printed when an
`Optional` type annotation with an unsupported contained type is
encountered. At present, the `Optional` is printed as-is, and
`Optional[T]` is syntatic sugar for `Union[T, None]`, so that is what
shows up in the error message and can be confusing. This commit modifies
the error message so that it prints `T` instead of `Union[T, None]`.

**Test Plan**
Continuous integration.

Example of old message:
```
AssertionError: Unsupported annotation typing.Union[typing.List, NoneType] could not be resolved.
```
Example of new message:
```
AssertionError: Unsupported annotation typing.Union[typing.List, NoneType] could not be resolved because typing.List could not be resolved.
```

**Fixes**
This commit fixes #42859.

Differential Revision: [D23490365](https://our.internmc.facebook.com/intern/diff/D23490365)

Commit:
[JIT] Improve error message for unsupported Optional types

**Summary**
This commit improves the error message that is printed when an
`Optional` type annotation with an unsupported contained type is
encountered. At present, the `Optional` is printed as-is, and
`Optional[T]` is syntatic sugar for `Union[T, None]`, so that is what
shows up in the error message and can be confusing. This commit modifies
the error message so that it prints `T` instead of `Union[T, None]`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #42859.

[ghstack-poisoned]

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -292,11 +292,14 @@ def try_ann_to_type(ann, loc):
         value = try_ann_to_type(ann.__args__[1], loc)
         return DictType(key, value)
     if is_optional(ann):
+        print(ann)
```
Remove?

SplitInfinity replied:
Embarrassing :'(

gmagogsfm commented on torch/jit/annotations.py:
```diff
@@ -292,11 +292,14 @@ def try_ann_to_type(ann, loc):
-        assert valid_type, "Unsupported annotation {} could not be resolved.".format(repr(ann))
+            contained = ann.__args__[1]
+        valid_type = try_ann_to_type(contained, loc)
+        assert valid_type, "Unsupported annotation {} could not be resolved.".format(repr(contained))
```
I think this is a good change to make message clearer.  Just one minor suggestion, what do you think about making this message a bit more verbose? Something like 
```
unsupported annotation Union[T, None] could not be resolved, because T could not be resolved.
```

gmagogsfm replied:
Yeah, good point, Optional[T] is better than Union[T, None]. I just wanted to make sure users don't get confused the other way : "why is JIT complaining about type being T when I clearly said it is Optional[T]"

SplitInfinity replied:
```
>>> from typing import Optional
>>> str(Optional[int])
'typing.Union[int, NoneType]'
```
I guess we can leave it as `Union` but add the extra phrasing you suggested.

Commit:
Update on "[JIT] Improve error message for unsupported Optional types"


**Summary**
This commit improves the error message that is printed when an
`Optional` type annotation with an unsupported contained type is
encountered. At present, the `Optional` is printed as-is, and
`Optional[T]` is syntatic sugar for `Union[T, None]`, so that is what
shows up in the error message and can be confusing. This commit modifies
the error message so that it prints `T` instead of `Union[T, None]`.

**Test Plan**
Continuous integration.

**Fixes**
This commit fixes #42859.

[ghstack-poisoned]

wconstab reviewed (approved):
Looks much better!
